### PR TITLE
PoV Distribution optimization

### LINKED ISSUE #1688: Optimize PoV Distribution: in-group gossip
As PoV Distribution currently works, we effectively gossip all PoVs for all scheduled paras to the entire network.

Now that we have validator discovery and connection APIs, we can limit the gossip of PoVs for a specific para to the validators of the group that they are assigned to.

That is, validators should attempt to connect to all other validators in their group and gossip PoVs for the para their group is assigned to only to those peers and not to the broader network. This should cut down on bandwidth substantially.

### PR DESCRIPTION
**Pull request by montekki**
It doesn't look like I've broken any existing tests or the only existing integration test so far. however I am most certain that some corner cases are not yet accounted for but for that I am needing fresh eyes and I will also be re-reading the code so marking at as ready for review now.

Fixes #1688

### CODE REVIEW

File: node/network/pov-distribution/src/lib.rs [hunk 1]
```diff
@@ -75,23 +89,45 @@ impl<C> Subsystem<C> for PoVDistribution
...
+
 	n_validators: usize,
 }
 
+/// Information kept about each PoV we attempt to fetch.
+struct PerFetchedPoV {
+	/// Channels expecting the data.
+	///
+	/// This may be an empty list, which indicates that we were once awaiting this PoV
+	/// but have received it already.
```
Comment (pepyakin) [inline node/network/pov-distribution/src/lib.rs:127]: I think this has some duplication with the `fetching` field above. Do you think we should deduplicate?

File: node/network/pov-distribution/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+	Subsystem(#[from] polkadot_subsystem::SubsystemError),
+	#[error(transparent)]
+	OneshotRecv(#[from] futures::channel::oneshot::Canceled),
+	#[error(transparent)]
+	Runtime(#[from] polkadot_subsystem::errors::RuntimeApiError),
+	#[error(transparent)]
+	ValidatorDiscovery(#[from] polkadot_node_subsystem_util::validator_discovery::Error),
+	#[error(transparent)]
+	Util(#[from] polkadot_node_subsystem_util::Error),
+}
```
Comment (pepyakin) [inline node/network/pov-distribution/src/error.rs:31]: Not sure about the team's policy about that, but many of those can be constructed from different places. Having, say, `OneshotRecv` at hand it would probably hard to tell where that error actually originated from.

File: node/network/pov-distribution/src/lib.rs [hunk 3]
```diff
@@ -540,50 +615,61 @@ impl PoVDistribution {
...
-							&mut ctx,
-							event,
-						).await?,
-				},
+			while let Poll::Ready(Some((_relay_parent, _validator_id, peer_id))) =
+				futures::poll!(state.connection_requests.next()) {
+					if let Err(err) = handle_validator_connected(
+						&mut state,
+						peer_id,
+					).await {
```
Comment (pepyakin) [inline node/network/pov-distribution/src/lib.rs:628]: Does this need to return a Result?

**Review by pepyakin — COMMENTED** [review]
A very shallow review because I am not very fluent in this area. Also, I've missed that this is a `inprogress`, sorry

File: node/network/pov-distribution/src/lib.rs (hunk 1 above)
Comment (montekki) [reply node/network/pov-distribution/src/lib.rs]: Yep already did that

File: node/network/pov-distribution/src/lib.rs [hunk 4]
```diff
@@ -540,50 +615,61 @@ impl PoVDistribution {
...
+							handle_network_update(
+								&mut state,
+								&mut ctx,
+								event,
+							).await?,
+					},
+				}
+			}
+
+			futures::pending!()
```
[the remark this answers is not in the document]
Comment (montekki) [reply node/network/pov-distribution/src/lib.rs:704]: In places like this we want to sort of manually re-create the logic of `select!` macro since using it in subsystem main loops usually runs into problems with borrowing. What we want is: poll everything that there is and when we've polled all things that are ready we need to tell the executor that we are pending until the next thing arrives. If you run the following code

```rust
use std::task::Poll;
use std::time::Duration;

use futures::{executor, StreamExt};

fn main() {
	executor::block_on(async move {
		let stream1 = futures::stream::repeat(()).then(|_| async move {
			futures_timer::Delay::new(Duration::from_secs(1)).await;
		});

		let stream2 = futures::stream::repeat(()).then(|_| async move {
			futures_timer::Delay::new(Duration::from_secs(1)).await;
		});

		futures::pin_mut!(stream1);
		futures::pin_mut!(stream2);

		loop {
			while let Poll::Ready(Some(_)) = futures::poll!(stream1.next()) {
				println!("stream1");
			}

			while let Poll::Ready(Some(_)) = futures::poll!(stream2.next()) {
				println!("stream2");
			}
		}
	});
}
```

you would find that without `futures::pending!()` it creates a busy loop since there is no place we back off.

File: node/network/pov-distribution/src/lib.rs [hunk 5]
```diff
@@ -254,50 +244,121 @@ async fn distribute_to_awaiting(
...
+
 /// Handles a `FetchPoV` message.
 async fn handle_fetch(
 	state: &mut State,
 	ctx: &mut impl SubsystemContext<Message = PoVDistributionMessage>,
 	relay_parent: Hash,
 	descriptor: CandidateDescriptor,
 	response_sender: oneshot::Sender<Arc<PoV>>,
-) -> SubsystemResult<()> {
+) -> crate::error::Result<()> {
```
[the remark this answers is not in the document]
Comment (montekki) [reply node/network/pov-distribution/src/lib.rs]: right

File: node/network/pov-distribution/src/lib.rs [hunk 6]
```diff
@@ -254,50 +244,121 @@ async fn distribute_to_awaiting(
...
+			if occupied.para_id == para_id {
+				return Ok(Some(((idx as u32).into(), cores.len())));
+			}
+		}
+	}
+
+	Ok(None)
+}
+
+/// Figure our a group of validators assigned to a given `ParaId`.
```
[the remark this answers is not in the document]
Comment (montekki) [reply node/network/pov-distribution/src/lib.rs:291]:
```suggestion
/// Figure out a group of validators assigned to a given `ParaId`.
```

File: parachain/test-parachains/adder/collator/tests/integration.rs [hunk 7]
```diff
@@ -19,7 +19,7 @@
 
 // If this test is failing, make sure to run all tests with the `real-overseer` feature being enabled.
 #[substrate_test_utils::test]
-#[cfg(feature = "real-overseer")]
+//#[cfg(feature = "real-overseer")]
```
Comment (montekki) [inline parachain/test-parachains/adder/collator/tests/integration.rs:22]: damn

```suggestion
#[cfg(feature = "real-overseer")]
```

**Comment by pepyakin** [comment]
One thing I realized, doesn't this require an impl guide update?

File: node/network/pov-distribution/src/lib.rs [hunk 8]
```diff
@@ -482,13 +568,20 @@ async fn handle_incoming_pov(
 	).await
 }
 
+/// Handles a newly connected validator in the context of some relay leaf.
+async fn handle_validator_connected(state: &mut State, peer_id: PeerId) {
+	if !state.peer_state.contains_key(&peer_id) {
+		state.peer_state.insert(peer_id.clone(), PeerState::default());
+	}
```
Comment (bkchr) [inline node/network/pov-distribution/src/lib.rs:575]:
```suggestion
	state.peer_state.entry(peer_id).or_default();
```

File: node/network/pov-distribution/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (bkchr) [reply node/network/pov-distribution/src/lib.rs:700]: Where do we run here into borrowing issues and can not use `select!`?

File: node/network/pov-distribution/src/lib.rs [hunk 9]
```diff
@@ -260,6 +270,73 @@ async fn distribute_to_awaiting(
...
+			warn!(
+				target: LOG_TARGET,
+				"Looks like no core is assigned to {:?} at {:?}", para_id, relay_parent,
+			);
+
+			return Ok(None);
+		}
+	};
+
+	Ok(determine_validators_for_core(ctx, core, num_cores, relay_parent).await?)
```
Comment (bkchr) [inline node/network/pov-distribution/src/lib.rs:337]:
```suggestion
	determine_validators_for_core(ctx, core, num_cores, relay_parent).await
```

File: node/network/pov-distribution/src/lib.rs [hunk 10]
```diff
@@ -260,6 +270,73 @@ async fn distribute_to_awaiting(
...
+	relay_parent: Hash,
+	para_id: ParaId,
+) -> error::Result<Option<Vec<ValidatorId>>> {
+	// Determine which core the para_id is assigned to.
+	let (core, num_cores) = match determine_core(ctx, para_id, relay_parent).await? {
+		Some(core) => core,
+		None => {
+			warn!(
+				target: LOG_TARGET,
+				"Looks like no core is assigned to {:?} at {:?}", para_id, relay_parent,
```
Comment (bkchr) [inline node/network/pov-distribution/src/lib.rs:330]:
```suggestion
				"Looks like no core is assigned to {:?} at {:?}",
				para_id,
				relay_parent,
```

File: node/network/pov-distribution/src/lib.rs [hunk 11]
```diff
@@ -22,22 +22,36 @@
...
+	request_availability_cores_ctx,
+	metrics::{self, prometheus},
+};
+use polkadot_node_network_protocol::{
+	v1 as protocol_v1, ReputationChange as Rep, NetworkBridgeEvent, PeerId, View,
+};
 
 use futures::prelude::*;
 use futures::channel::oneshot;
+use log::warn;
```
Comment (bkchr) [inline node/network/pov-distribution/src/lib.rs:48]: I thought we are using now `tracing`?

File: node/network/pov-distribution/Cargo.toml [hunk 12]
```diff
@@ -6,14 +6,22 @@ edition = "2018"
 
 [dependencies]
 futures = "0.3.8"
+log = "0.4.11"
```
Comment (bkchr) [inline node/network/pov-distribution/Cargo.toml:9]: Same

File: node/network/pov-distribution/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (montekki) [reply node/network/pov-distribution/src/lib.rs]: Thas is not always the case and here I changed the code to usage of `select_biased`.

File: node/network/pov-distribution/src/lib.rs [hunk 13]
```diff
@@ -561,45 +652,59 @@ impl PoVDistribution {
...
+			futures::select_biased! {
+				v = state.connection_requests.next().fuse() => {
+					match v {
+						Some((_relay_parent, _validator_id, peer_id)) => {
+							handle_validator_connected(&mut state, peer_id);
+						}
+						None => break,
+					}
+				}
+				v = ctx.recv().fuse() => {
```
Comment (pepyakin) [inline node/network/pov-distribution/src/lib.rs:672]: Here and above a call to `.fuse()` happens while `select_biased` is enclosed by a loop.

The docs for [`select_biased`](https://docs.rs/futures/0.3.8/futures/macro.select_biased.html) state the following:

> Futures and streams which are not already fused can be fused using the .fuse() method. Note, though, that fusing a future or stream directly in the call to select_biased! will not be enough to prevent it from being polled after completion if the select_biased! call is in a loop, so when select_biased!ing in a loop, users should take care to fuse() outside of the loop.

This seems to apply to this particular case, right?

Comment (montekki) [reply node/network/pov-distribution/src/lib.rs:672]: Long story short this happens. 

What `Fuse` is is this

```rust
pub struct Fuse<Fut>(#[pin] Option<Fut>);
```

(more or less the same for `Stream`). The idea is that we want a protection against polling a Future or a Stream that have finished.

The `connection_requests` field is a `Stream` so we would normally `.fuse()` the stream as the docs say before the loop and just be done with it. The problem with this is that `.fuse()` consumes `self` and we would have to move the `connection_requests` out of the `state` and `.fuse()` it. If that is done the code would no longer be able to update `connection_requests` as needed. So another approach is taken. Calling `.next()` on the `Stream` is essentially a `Future` that may be `await`ed or what not so we just want to create such a `Future` and `.fuse()` it here. It will be anyway thrown away after it becomes ready.

The same thing goes with the `recv()` which is also a `Future`. but this case is just a part of a previously described one. 

I will add a comment although I believe that this approach is not novel in the codebase: at least `collation-generation` and `subsystem-util` use the same one.

File: node/network/pov-distribution/src/error.rs [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
```
Comment (rphmeier) [inline node/network/pov-distribution/src/error.rs:16]: missing crate docs

File: node/network/pov-distribution/src/lib.rs [hunk 15]
```diff
@@ -289,7 +366,21 @@ async fn handle_fetch(
 				return Ok(());
 			}
 			Entry::Vacant(e) => {
-				e.insert(vec![response_sender]);
+				if let Some(relevant_validators) = determine_relevant_validators(
+					ctx,
+					relay_parent,
+					descriptor.para_id,
+				).await? {
+					let new_connection_request = validator_discovery::connect_to_validators(
```
Comment (rphmeier) [inline node/network/pov-distribution/src/lib.rs:374]: Could it be too heavy to constantly call `connect_to_validators`? Does this call typically return instantly if we're already connected to those validators?

Comment (rphmeier) [reply node/network/pov-distribution/src/lib.rs:374]: I think our expectation is that we'll call this function N times per relay-parent, where N is the number of other validators in our group.

File: node/network/pov-distribution/src/lib.rs [hunk 16]
```diff
@@ -482,13 +568,18 @@ async fn handle_incoming_pov(
 	).await
 }
 
+/// Handles a newly connected validator in the context of some relay leaf.
+fn handle_validator_connected(state: &mut State, peer_id: PeerId) {
+	state.peer_state.entry(peer_id).or_default();
```
Comment (rphmeier) [inline node/network/pov-distribution/src/lib.rs:573]: It doesn't seem to do anything differently than the regular `PeerConnected` logic?

File: node/network/pov-distribution/src/lib.rs [hunk 17]
```diff
@@ -289,7 +366,21 @@ async fn handle_fetch(
...
+					relay_parent,
+					descriptor.para_id,
+				).await? {
+					let new_connection_request = validator_discovery::connect_to_validators(
+						ctx,
+						relay_parent,
+						relevant_validators.clone(),
+					).await?;
+
+					state.connection_requests.put(relay_parent, new_connection_request);
```
Comment (rphmeier) [inline node/network/pov-distribution/src/lib.rs:380]: I suppose we only need one connection request per `(relay_parent, para_id)` pair, in actuality. And even this is possibly overkill.

File: node/network/pov-distribution/src/lib.rs (hunk 15 above)
Comment (ordian) [reply node/network/pov-distribution/src/lib.rs:374]: connect_to_validators returns a `request` which is a stream that can be polled to retrieve the ids as we discover them (and instantly if already known), so it shouldn't be too heavy latency-wise unless we have too many revoked requests to process.

**Review by rphmeier — COMMENTED** [review]
I see the changes where we connect to the peers in our group about the PoV, but not the changes where we only notify and request from peers in that group. We will need to change the logic for computing `peers_to_send` also.

File: node/network/pov-distribution/src/lib.rs (hunk 15 above)
Comment (rphmeier) [reply node/network/pov-distribution/src/lib.rs:374]: OK, that's workable then!

**Comment by montekki** [comment]
My current understanding of the protocol is this:

The first pathway is `FetchPoV`. The actions there are:
 1) determine relevant validators
 2) for every validator as they connect issue a relevant `Awaiting` message.

The second pathway is `DistributePoV` and it will do this:
 1) Store the known `PoV`.
 2) For every peer that we have previously received `Awaiting` issue a `SendPoV` message.

The path for networking messages: just handle `SendPoV` and `Awaiting` messages.

In other words, the only proactive networking actions with connecting to relevant validators are taken on fetches, and we can assume that on distributing path we only receive `Awaiting` messages from the relevant peers.

Or am I missing something?

File: node/network/pov-distribution/src/lib.rs (hunk 13 above)
Comment (bkchr) [reply node/network/pov-distribution/src/lib.rs:672]: You can just implement `FusedStream` for `ConnectionRequests` with `is_terminated` returning always `false`.

File: node/network/pov-distribution/src/lib.rs [hunk 18]
```diff
@@ -561,45 +652,59 @@ impl PoVDistribution {
...
+							PoVDistributionMessage::NetworkBridgeUpdateV1(event) =>
+								handle_network_update(
+									&mut state,
+									&mut ctx,
+									event,
+								).await?,
+						}
+					}
+				}
+				complete => break,
```
Comment (bkchr) [inline node/network/pov-distribution/src/lib.rs:703]: Why?

**Comment by rphmeier** [comment]
> My current understanding of the protocol is this:
> 
> The first pathway is `FetchPoV`. The actions there are:
> 
>     1. determine relevant validators
> 
>     2. for every validator as they connect issue a relevant `Awaiting` message.
> 
> 
> The second pathway is `DistributePoV` and it will do this:
> 
>     1. Store the known `PoV`.
> 
>     2. For every peer that we have previously received `Awaiting` issue a `SendPoV` message.
> 
> 
> The path for networking messages: just handle `SendPoV` and `Awaiting` messages.
> 
> In other words, the only proactive networking actions with connecting to relevant validators are taken on fetches, and we can assume that on distributing path we only receive `Awaiting` messages from the relevant peers.
> 
> Or am I missing something?

That sounds right in principle, but I don't really understand what the code is doing. 

Re point 2, the code here doesn't dispatch any `awaiting` message: https://github.com/paritytech/polkadot/pull/1990/files#diff-578deb92139a25c42178fc12e8a62dde2b5b4dfe6fe0ab30b698e8caead3835fR573 . It initializes the `PeerState` if there wasn't one already, but this is something that the regular `PeerConnected` event does already, so as far as I see this code is currently redundant. I would expect it to somehow tag the `PeerState` in such a way as to inform the logic of `notify_all_we_are_awaiting` and `notify_one_we_are_awaiting_many`.

 And the call to `notify_all_we_are_awaiting` that used to be in `handle_fetch` was now removed: https://github.com/paritytech/polkadot/pull/1990/files#diff-578deb92139a25c42178fc12e8a62dde2b5b4dfe6fe0ab30b698e8caead3835fL306 . AFAIK nothing calls that code anymore.

Also, the call to `notify_one_we_are_awaiting_many` on view change has not been altered, so we'll still inform peers about everything we're awaiting, regardless of group: https://github.com/paritytech/polkadot/pull/1990/files#diff-578deb92139a25c42178fc12e8a62dde2b5b4dfe6fe0ab30b698e8caead3835fR605 .

**Comment by montekki** [comment]
>  I suggest to reinstate this call, but I'm not sure why you removed it in the first place.

Ah yes I agree. For some time I was thinking that we would re-receive the `PeerViewUpdate` message anyway so we would hit that path. Expanded a test to this situation.

Also looks like the attributes on functions hide `unused` compiler warnings

File: node/network/pov-distribution/src/lib.rs (hunk 16 above)
Comment (montekki) [reply node/network/pov-distribution/src/lib.rs:580]: I've unified them, yes

File: node/network/pov-distribution/src/lib.rs (hunk 17 above)
Comment (montekki) [reply node/network/pov-distribution/src/lib.rs]: agree

File: node/network/pov-distribution/src/lib.rs (hunk 13 above)
Comment (ordian) [reply node/network/pov-distribution/src/lib.rs:676]: Implementing `FusedStream` doesn't help.

As mentioned above, once you `.fuse()` a stream, you get back a stream-only interface. Whereas we want to call other methods on it, like `.put(..)` and `.remove(..)`.

File: node/network/pov-distribution/src/lib.rs [hunk 19]
```diff
@@ -121,45 +146,24 @@ async fn handle_signal(
...
-
-						// Not adding bookkeeping here might make us behave funny, but we
-						// shouldn't take down the node on spurious runtime API errors.
-						//
-						// and this is "behave funny" as in be bad at our job, but not in any
-						// slashable or security-related way.
-						continue;
-					}
-				};
+				let n_validators = request_validators_ctx(relay_parent.clone(), ctx).await?.await??.len();
```
Comment (ordian) [inline node/network/pov-distribution/src/lib.rs:156]: we previously `continue`d on runtime error, what was the rationale for changing it to bail?

File: node/network/pov-distribution/src/lib.rs [hunk 20]
```diff
@@ -557,45 +656,58 @@ impl PoVDistribution {
...
+						}
+						FromOverseer::Communication { msg } => match msg {
+							PoVDistributionMessage::FetchPoV(relay_parent, descriptor, response_sender) =>
+								handle_fetch(
+									&mut state,
+									&mut ctx,
+									relay_parent,
+									descriptor,
+									response_sender,
+								).await?,
```
Comment (ordian) [inline node/network/pov-distribution/src/lib.rs:689]: we should not early-exit on error here, the error here is a runtime error IIUC

File: node/network/pov-distribution/src/lib.rs [hunk 21]
```diff
@@ -557,45 +656,58 @@ impl PoVDistribution {
...
 		self,
 		mut ctx: impl SubsystemContext<Message = PoVDistributionMessage>,
-	) -> SubsystemResult<()> {
-		let mut state = State {
-			relay_parent_state: HashMap::new(),
-			peer_state: HashMap::new(),
-			our_view: View(Vec::new()),
-			metrics: self.metrics,
-		};
+	) -> error::Result<()> {
```
Comment (ordian) [inline node/network/pov-distribution/src/lib.rs:659]: I suggest to revert the return type back to `SubsystemResult<()>` or even remove it.
The errors should not propagate.

**Review by ordian — COMMENTED** [review]
Looks good modulo error handling.

File: node/network/pov-distribution/src/lib.rs (hunk 13 above)
Comment (montekki) [reply node/network/pov-distribution/src/lib.rs:683]: It does.

File: node/network/pov-distribution/src/lib.rs (hunk 19 above)
Comment (montekki) [reply node/network/pov-distribution/src/lib.rs]: The other call about `request_validators_ctx` should have been a `continue` not this one, i've fixed that.

**Review by rphmeier — APPROVED** [review]

**Comment by montekki** [comment]
bot merge
